### ISSUE TITLE
RFC: Improved connectBackboneToReact and BackboneProvider

### ISSUE THREAD

**Issue by hswolff** [mongodb-js/connect-backbone-to-react#2]
##  How do I RFC?

* Read the [Problem](#1-problem) and [Goals](#2-goals), then the proposed solution. 
* If you like the [Proposed Solution](#3-proposed-solution), give a :+1: or an LGTM.
* If you dislike the proposed solution, check out the [alternate solutions](#4-alternate-solutions).
    * If you like one of them, propose one of them and explain why.
    * If you sort of like one of them, propose a change to it.
    * If you dislike all of them, propose a new alternate solution.
* If you'd rather this be something we don't regulate, voice that.

Statuses for RFCs are:
* Discussion
* Adopted
* Not Regulated

## Contents

* [1. Problem](#1-problem)
* [2. Goals](#2-goals)
* [3. Proposed Solution](#3-proposed-solution)
* [4. Alternate Solutions](#4-alternate-solutions)

## 1. Problem

In order to use React Components within Backbone we need to accomplish two things:

1. Be able to render a React Component within Backbone. (This is addressed in a PR).

  After the PR is merged we will be able to render a React Component within a Marionette View like this:

  ```javascript
  class Greeting extends Component {
    render() {
      return (
        <div>
          Hello!
        </div>
      );
    }
  }

  const MyView = mixInto(Marionette.ItemView)(
      viewWithReact,
  ).extend({
      template: () => `
        <div>
            <header>
                <h1>Welcome!</h1>
            </header>

            <div class="react"></div>
        </div>
      `,

      initialize: function(options) {
      
        this.registerReactComponent(
          <Greeting />,
          '.react'
        );
      }
  });

  (new MyView()).render();
  // Rendered!
  ```

2. Be able to share data stored within Backbone Models and Collections between React Components and Backbone Views.

  Consider this case:

  ```javascript
  class Greeting extends Component {
    render() {
      return (
        <div>
          Hello {this.props.name}!
        </div>
      );
    }
  }

  const MyView = mixInto(Marionette.ItemView)(
      viewWithReact,
  ).extend({
      template: () => `
        <div>
            <header>
                <h1>Welcome!</h1>
            </header>

            <div class="react"></div>
        </div>
      `,

      initialize: function(options) {
        const model = new Backbone.Model({
          name: 'Harry',
        });

        // We could do this:

        this.registerReactComponent(
          <Greeting name={model.get('name')} />,
          '.react'
        );

        // However if we then do:

        model.set('name', 'The Loud One');

        // Then the React Component is out of sync with the data.

        // This can be fixed via:

        model.on('change:name', () => {
          this.registerReactComponent(
            <Greeting name={model.get('name')} />,
            '.react'
          );

          this.render();
        });

        // But that quickly gets cumbersome.

        // Alternatively we could fix it like:
        this.registerReactComponent(
          <Greeting model={model} />,
          '.react'
        );

        // And within the Greeting component set up our event handlers directly,
        // but again that gets cumbersome to repeat that boilerplate every time.

        // Additionally a React Component should not have a direct dependence on a Model.
        // We should aim to decouple.
      }
  });

  (new MyView()).render();
  ```

This RFC addresses the second issue.

## 2. Goals

1. Retain one true source of truth. That being, the data stored within the Models and Collections is what all UI should use to render.
2. Create an abstraction between Backbone Models/Collections and React Components. This is to prevent a React Component from being directly dependent on a Model/Collection.
3. Keep the UI in sync with the data in the Models/Collections. Such that whenever the Model/Collection changes the React Component re-renders.
4. Create an API that is as frictionless as possible. This is to encourage usage.

## 3. Proposed Solution

An initial solution has been created via [connect-backbone-to-react](https://github.com/mongodb-js/connect-backbone-to-react) package, however the API has room for improvement before we mark it as `1.0.0`.

This solution uses the [Higher-Order Components](https://facebook.github.io/react/docs/higher-order-components.html) (HoCs) pattern. This is the solution the React community has moved to after finding mixins hard to maintain. The end goal is the same as mixins, sharing functionality, however HoCs are easier to reason about. Refer to that document for further explanation.

`connectBackboneToReact` creates a HoC. The HoC returned:

* Manages creating Model/Collection eventListeners so that when the model changes, the enclosed component is re-rendered.
* Handles cleaning up eventListeners when the component is unmounted.
* Maps Models/Collections to `props` that are given to your component.
* Passes any additional props given to the HoC to the WrappedComponent. This allows you to still pass `props` that change non-data related behaviors, such as `showPlaceholder={false}`.

Ultimately we have four inputs required to correctly create our HoC:

- `modelsMap` - this defines the map of Models/Collections to local names.
  - Example: `{ user: userModel, settings: settingsCollection }`

- `mapModelsToProps` - this is function that maps the values from `modelsMap` to `props` that will be given to our WrappedComponent.
  - Example:
    ```javascript
    function({ user, settings }) {
      return {
        name: user.get('name'),
        changeName: (newName) => user.set('name', newName)
      };
    };
    ```

- `options` - general options object to manipulate how the wrapper behaves.
  - `options.debounce`: Should our event handler function be wrapped in a debounce function. By default it's _false_.
  - `options.events`: Custom map of events our HoC should subscribe to. By default the _all_ event is used for each Model/Collection.
  - `options.modelTypes`: Define what Model/Collection types the HoC expects to be given in the `modelsMap`. This is patterned off of [React's PropTypes](https://facebook.github.io/react/docs/typechecking-with-proptypes.html).

-  `WrappedComponent` - our own React Component that is going to be wrapped.

The full usage would look like:

```javascript
// ComponentA.js
const { connectBackboneToReact } = require('connect-backbone-to-react');

class ComponentA extends Component {
  render() {
    return (
      <div>
        Hello {this.props.name}!
        <div
          onClick={() => {
            this.props.setName(Math.random());
          }}
      </div>
    );
  }
}

const mapModelsToProps = ({ user }) => {
  return {
    name: user.get('name'),
    setName: (newName) => user.set('name', newName)
  };
},
const options = {
  debounce: true,
  events: {
    user: ['change', 'reset'],
  },
  modelTypes: {
    user: Backbone.Model,
  },
};

module.exports = connectBackboneToReact(
  mapModelsToProps,  
  options
)(ComponentA);

// MyView.js

// ComponentA is the connected HoC.
const ComponentA = require('./ComponentA.js');

const MyView = mixInto(Marionette.ItemView)(
  viewWithReact,
).extend({
  template,

  initialize: function(options) {
    const model = new Backbone.Model({
      name: 'Harry',
    });

    const modelsMap = {
      user: model,
    };

    this.registerReactComponent(
      <ComponentA models={modelsMap} />,
      '.react'
    );
  }
});
```

---

That covers the basic use case of rendering one connected React Component. However as we incrementally migrate more Backbone Views to React we'll end up with a tree of React Components. So what happens if we have a tree of connected React Components?

```javascript
class ComponentA extends Component {
  render() {
    return (
      <div>
        <ConnectedComponentB>
          <ConnectedComponentC />
        </ConnectedComponentB>
      </div>
    );
  }
}
```

We shouldn't pass that `modelsMap` object from one component to another. Instead we can take inspiration from react-redux's [Provider component](https://github.com/reactjs/react-redux/blob/master/docs/api.md#provider-store).

This allows us to share the `modelsMap` via [React's Context feature](https://facebook.github.io/react/docs/context.html).

Usage would look like:

```javascript
// ComponentA is the connected HoC.
const ComponentA = require('./ComponentA.js');
const { BackboneProvider } = require('connect-backbone-to-react');

const MyView = mixInto(Marionette.ItemView)(
  viewWithReact,
).extend({
  template,

  initialize: function(options) {
    const model = new Backbone.Model({
      name: 'Harry',
    });

    const modelsMap = {
      user: model,
    };

    this.registerReactComponent(
      <BackboneProvider models={modelsMap}>
        <ComponentA />
      </BackboneProvider>,
      '.react'
    );
  }
});
```

`BackboneProvider` shares the `modelsMap` via `context`. Every HoC connected to Backbone would then retrieve the `modelsMap` via `context`. 

### What should the `props` name be that the `modelsMap` is given to?

Right now you pass in `modelsMap` through the `models` prop:

```javascript
<ComponentA models={modelsMap} />
```

Alternatively: 

```javascript
<ComponentA data={modelsMap} />
```

### Where should we put `modelTypes`?

We have the proposed solution, here's an alternative which mirrors the style that `PropTypes` are added to React components.

```javascript
const GreetingConnected = connectBackboneToReact(
  mapModelsToProps,
  options,
)(Greeting);

GreetingConnected.modelTypes = {
  user: instanceOf(UserModel),
  allUsers: instanceOf(UserCollection),
};
```

**Comment by jrbalsano** [comment]
Two questions, for clarification:
* What does it look like if we don't use the provider to get the models from `ComponentA` to `ComponentC`?
* What does it look like if we need to pass some sort of non-backbone property to `ComponentA`?

**Comment by hswolff** [comment]
> What does it look like if we don't use the provider to get the models from ComponentA to ComponentC?

So ideally that should never be done. As that breaks the abstraction between Backbone and React.

However if you truly wanted to do that, let's say `ComponentA` is the HoC. You would probably have to pass the `modelsMap` object straight through the `mapModelsToProps` and then give that to `ComponentC`.

```javascript
class ComponentA extends Component {
  render() {
    return (
      <div>
        <ComponentC models={this.props.modelsMap} />
      </div>
    );
  }
}

const mapModelsToProps = ({ user }) => {
  return {
    name: user.get('name'),
    modelsMap: { user }
  };
},
module.exports = connectBackboneToReact(
  mapModelsToProps,  
)(ComponentA);
```

> What does it look like if we need to pass some sort of non-backbone property to ComponentA?

You can do that just as you would any other React component.

```javascript
const MyView = mixInto(Marionette.ItemView)(
  viewWithReact,
).extend({
  template,

  initialize: function(options) {
    const model = new Backbone.Model({
      name: 'Harry',
    });

    const modelsMap = {
      user: model,
    };

    this.registerReactComponent(
      <ComponentA models={modelsMap} hideHeader={true} />,
      '.react'
    );
  }
});
```

**Comment by jjgonecrypto** [comment]
Thanks for the ultra-clear RFC @hswolff. I initially thought that contexts make the most sense, though I don't love that it's: 
>  It is an experimental API and it is likely to break in future releases of React.

But looking deeper, it seems every parent has to contain both `getChildContext` and `childContextTypes`, which seems onerous, compared to just passing in props all the way down. Considering the API is experimental, and it ends up being more code, would we really prefer to use contexts?

**Comment by hswolff** [comment]
Ah, so `childContextTypes` will be handled by the HoC returned by the `connectBackboneToReact` function. That HoC will be responsible for grabbing the `modelsMap` off of the `contextTypes` object.

**Comment by hswolff** [comment]
This RFC has been implemented in #3. Happy to update with any feedback.
